dttools/src/mq.c

```diff
@@ -0,0 +1,716 @@
+	return out;
+}
+#endif
+
+static int ppoll_compat(struct pollfd fds[], nfds_t nfds, int stoptime) {
+	assert(fds);
+	sigset_t mask;
+	sigemptyset(&mask);
+	int timeout = stoptime - time(NULL);
+	if (timeout < 0) return 0;
+
+#ifdef HAS_PPOLL
+	struct timespec s;
+	s.tv_nsec = 0;
+	s.tv_sec = timeout;
+	return ppoll(fds, nfds, &s, &mask);
+#else
+	sigset_t origmask;
+	sigprocmask(SIG_SETMASK, &mask, &origmask);
+	int rc = poll(fds, nfds, timeout*1000);
+	int saved_errno = errno;
+	sigprocmask(SIG_SETMASK, &origmask, NULL);
+	errno = saved_errno;
+	return rc;
+#endif
+}
+
+
+static bool errno_is_temporary(void) {
+	if (errno == EINTR || errno == EWOULDBLOCK || errno == EAGAIN || errno == EINPROGRESS || errno == EALREADY || errno == EISCONN) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+static struct mq_msg *msg_create (enum mq_msg_type type) {
+	// sanity check
+	assert(HDR_SIZE == 16);
+	struct mq_msg *out = xxcalloc(1, sizeof(struct mq_msg));
+	buffer_init(&out->buf);
+	buffer_abortonfailure(&out->buf, true);
+	memcpy(out->magic, HDR_MAGIC, sizeof(out->magic));
+	out->type = type;
+	return out;
+}
+
+static void mq_die(struct mq *mq, int err) {
+	assert(mq);
+	mq->state = MQ_SOCKET_ERROR;
+	mq->err = err;
+
+	mq_close(mq->acc);
+	mq_msg_delete(mq->send_buf);
+	mq_msg_delete(mq->recv_buf);
+	mq_msg_delete(mq->recv);
+
+	struct list_cursor *cur = list_cursor_create(mq->send);
+	list_seek(cur, 0);
+	for (struct mq_msg *msg; list_get(cur, (void **) &msg); list_next(cur)) {
+		mq_msg_delete(msg);
+	}
+	list_cursor_destroy(cur);
+
+	if (mq->poll_group) {
+		set_remove(mq->poll_group->acceptable, mq);
+		set_remove(mq->poll_group->readable, mq);
+		if (err == 0) {
+			set_remove(mq->poll_group->error, mq);
+		} else {
+			set_insert(mq->poll_group->error, mq);
+		}
+	}
+}
+
+static void delete_msg(struct mq_msg *msg) {
+	if (!msg) return;
+	buffer_free(&msg->buf);
+	jx_delete(msg->j);
+	free(msg);
+}
+
+static struct mq *mq_create(enum mq_socket state, struct link *link) {
+	struct mq *out = xxcalloc(1, sizeof(*out));
+	out->send = list_create();
+	out->state = state;
+	out->link = link;
+	return out;
+}
+
+void mq_close(struct mq *mq) {
+	if (!mq) return;
+
+	mq_die(mq, 0);
+	if (mq->poll_group) {
+		itable_remove(mq->poll_group->members, (uintptr_t) mq);
+	}
+	link_close(mq->link);
+	list_delete(mq->send);
+	free(mq);
+}
+
+int mq_geterror(struct mq *mq) {
+	assert(mq);
+	if (mq->state != MQ_SOCKET_ERROR) {
+		return 0;
+	} else {
+		return mq->err;
+	}
+}
+
+void mq_msg_delete(struct mq_msg *msg) {
+	if (!msg) return;
+	// once blobs are implemented, check for on-disk stuff to delete
+	delete_msg(msg);
+}
+
+static int validate_header(struct mq_msg *msg) {
+	assert(msg);
+	errno = EBADF;
+	if (memcmp(msg->magic, HDR_MAGIC, sizeof(msg->magic))) {
+		return -1;
+	}
+	switch (msg->type) {
+		case MQ_MSG_BUFFER:
+		case MQ_MSG_JSON:
+			break;
+		default:
+			return -1;
+	}
+	return 0;
+}
+
+static int unpack_msg(struct mq_msg *msg) {
```

Comment:
As above, we may want to leave any jx out of this library, and add the jx conversion in DS proper.